java-support/src/main/java/io/cloudstate/javasupport/KeyValue.java

```diff
@@ -0,0 +1,200 @@
+      return this._reader;
+    }
+
+    public final Function<T, ByteString> writer() {
+      return this._writer;
+    }
+
+    @Override
+    public final boolean equals(final Object that) {
+      if (this == that) return true;
+      else if (that instanceof Key) return ((Key<T>) that).name().equals(this.name());
+      else return false;
+    }
+
+    @Override
+    public final int hashCode() {
+      return 403 + name().hashCode();
+    }
+
+    @Override
+    public final int compareTo(final Key<?> other) {
+      return name().compareTo(other.name());
+    }
+  }
+
+  public static final <T> Key<T> keyOf(
+      final String name,
+      final Function<ByteString, T> reader,
+      final Function<T, ByteString> writer) {
+    return new Key<T>(
+        requireNonNull(name, "Key name cannot be null"),
+        requireNonNull(reader, "Key reader cannot be null"),
+        requireNonNull(writer, "Key writer cannot be null"));
+  }
+
+  public static final class Map {
+    /*
+      TODO document invariants
+    */
+    private final java.util.Map<String, ByteString> unparsed;
+    private final java.util.Map<Key<?>, Object> updated = new java.util.TreeMap<Key<?>, Object>();
+    private final java.util.Set<String> removed = new java.util.TreeSet<String>();
+
+    Map(final java.util.Map<String, ByteString> initial) {
+      this.unparsed = requireNonNull(initial, "Map initial values must not be null");
+    }
+
+    public final <T> Optional<T> get(final Key<T> key) {
+      final T value = (T) updated.get(key);
+      if (value != null) {
+        return Optional.of(value);
+      } else {
+        final ByteString bytes = unparsed.get(key.name());
+        if (bytes != null) {
+          final T parsed = key.reader().apply(bytes);
+          requireNonNull(parsed, "Key reader not allowed to read `null`");
+          updated.put((Key<Object>) key, parsed);
+          return Optional.of(parsed);
+        } else {
+          return Optional.empty();
+        }
+      }
+    }
+
+    public final <T> void set(final Key<T> key, final T value) {
+      requireNonNull(key, "Map key must not be null");
+      requireNonNull(value, "Map value must not be null");
+      updated.put(key, value);
+      unparsed.remove(key.name());
+      removed.remove(key.name());
+    }
+
+    public final <T> boolean remove(final Key<T> key) {
+      requireNonNull(key, "Map key must not be null");
+      if (!removed.contains(key.name())
+          && (updated.remove(key) != null
+              | unparsed.remove(key.name()) != null)) { // Yes, you read that strict-or right
+        removed.add(key.name());
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    final KVEntityOrBuilder toProto() {
+      final KVEntity.Builder builder = KVEntity.newBuilder();
+      unparsed.forEach(
+          (k, v) -> {
+            builder.putEntries(k, com.google.protobuf.ByteString.copyFrom(v.asByteBuffer()));
```

Comment:
@viktorklang 
This is the serialization for the user choosen for his values in the KVS, right?
Does this conflict or is it compatible with the [stable serialization rules by Cloudstate](https://cloudstate.io/docs/user/lang/java/serialization.html)? Is this relevant here too? I think about how, if done, a separate/different user lib, language wise, would parse the values back is they're are marshalled as ByteStrings here.

Reply:
Yeah, this needs work. I am not sure how to integrate it though—I'd need to go through AnySupport I think.

Reply:
Could be a good example how to use AnySupport.

Reply:
Indeed. I'll have a look at what that could look like.